service/matching/taskWriter.go

```diff
@@ -80,23 +82,30 @@ func newTaskWriter(
 		config:       tlMgr.config,
 		taskQueueID:  tlMgr.taskQueueID,
 		appendCh:     make(chan *writeTaskRequest, tlMgr.config.OutstandingTaskAppendsThreshold()),
+		taskIDBlock:  noTaskIDs,
+		maxReadLevel: noTaskIDs.start - 1,
 		logger:       tlMgr.logger,
 		shutdownChan: make(chan struct{}),
 	}
 }
 
-func (w *taskWriter) Start(initialBlock taskIDBlock) {
+func (w *taskWriter) Start() {
 	if !atomic.CompareAndSwapInt32(
 		&w.status,
 		common.DaemonStatusInitialized,
 		common.DaemonStatusStarted,
 	) {
 		return
 	}
-
-	w.taskIDBlock = initialBlock
-	w.maxReadLevel = initialBlock.start - 1
-	go w.taskWriterLoop()
+	go func() {
+		if state, err := w.tlMgr.renewLeaseWithRetry(suppressFatalSignal); err == nil {
+			w.taskIDBlock = w.tlMgr.rangeIDToTaskIDBlock(state.rangeID)
+			atomic.StoreInt64(&w.maxReadLevel, w.taskIDBlock.start-1)
+			w.tlMgr.taskAckManager.setAckLevel(state.ackLevel)
+			w.tlMgr.taskReader.Signal()
+		}
```

Comment:
I think we don't nee this block, right? The taskWriterLoop() will check if it runs out of the id block and try to renew there.

Reply:
The problem becomes that if there are tasks buffered in the db, the taskReader won't start (or more accurately: won't start from the right place) until we assign taskWriter.maxReadLevel and call taskReader.Signal().

This would happen as a side-effect of producer-side activity (i.e. a side-effect of Add*Task) but in the absence of traffic the taskReader will stall. I could see this _maybe_ happening on a real system but it _definitely_ happens in unit tests.